snapshot/overlay/overlay.go

```diff
@@ -9,33 +9,98 @@ import (
 	"sync"
 
 	"github.com/docker/containerd"
+	"github.com/docker/containerd/snapshot"
 	digest "github.com/opencontainers/go-digest"
+	"github.com/pkg/errors"
 )
 
-type Driver struct {
+type Snapshotter struct {
 	root  string
-	cache *cache
+	links *cache
 }
 
-func NewDriver(root string) (*Driver, error) {
+func NewSnapshotter(root string) (*Snapshotter, error) {
 	if err := os.MkdirAll(root, 0700); err != nil {
 		return nil, err
 	}
 	for _, p := range []string{
-		"snapshots",
-		"active",
+		"committed", // committed snapshots
+		"active",    // active snapshots
+		"index",     // snapshots by hashed name
 	} {
 		if err := os.MkdirAll(filepath.Join(root, p), 0700); err != nil {
 			return nil, err
 		}
 	}
-	return &Driver{
+	return &Snapshotter{
 		root:  root,
-		cache: newCache(),
+		links: newCache(),
 	}, nil
 }
 
-func (o *Driver) Prepare(key, parent string) ([]containerd.Mount, error) {
+// Stat returns the info for an active or committed snapshot by name or
+// key.
+//
+// Should be used for parent resolution, existence checks and to discern
+// the kind of snapshot.
+func (o *Snapshotter) Stat(key string) (snapshot.Info, error) {
+	path, err := o.links.get(filepath.Join(o.root, "index", hash(key)))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return snapshot.Info{}, err
+		}
+
+		return snapshot.Info{}, errors.Errorf("snapshot %v not found", key)
+	}
+
+	// TODO(stevvooe): We don't confirm the name to avoid the lookup cost.
+	return o.stat(path)
+}
+
+func (o *Snapshotter) stat(path string) (snapshot.Info, error) {
+	ppath, err := o.links.get(filepath.Join(path, "parent"))
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return snapshot.Info{}, err
+		}
+
+		// no parent
+	}
+
+	kp, err := ioutil.ReadFile(filepath.Join(path, "name"))
+	if err != nil {
+		return snapshot.Info{}, err
+	}
+
+	var parent string
+	if ppath != "" {
+		p, err := ioutil.ReadFile(filepath.Join(ppath, "name"))
+		if err != nil {
+			return snapshot.Info{}, err
+		}
+		parent = string(p)
+	}
+
+	ro := true
+	kind := snapshot.KindCommitted
+	if strings.HasPrefix(path, filepath.Join(o.root, "active")) {
```

Comment:
This conditional looks pretty much the same as the `btrfs` implementation.  Do you plan to factor out determining the `kind` to a common implementation in the future?